test/sandbox/test/view_component_system_test.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require "test_helper"
+
+class ViewComponentSystemTest < ViewComponent::SystemTestCase
+  driven_by :cuprite
+
+  def test_simple_js_interaction_in_browser_without_layout
+    with_rendered_component_path(SimpleJavascriptInteractionWithJsIncludedComponent.new) do |page|
```

Comment:
Can you add a test that uses slots as well as a test that passes arguments to the component?

Reply:
@joelhawksley I added a basic test for the passing arguments. But I can't seem to figure out how to create a component with slots that I can pass into the helper. I see in test/sandbox/test/slotable_v2_test.rb that `render_inline` accepts and provides a block where slots can be added but not sure exactly how to handle it without using `render_inline`

Wondering if you can point me in the right direction here... or if we need to improve the helper to handle this use case?

Reply:
The added test looks good! I do think we need one to show that slots work (or that they don't, and that they should).

Something like:

```ruby
c = SlotsV2Component.new(classes: "mt-4") do |component|
  component.title do
    "This is my title!"
  end
with_rendered_component_path(c) do |path|
  visit path
  assert_selector(".title", text: "This is my title!")
end
```

Reply:
@joelhawksley hey there, I tried that out and it doesn't seem to be working. Here is what I tried;
```
  def test_components_with_slots
    c = SlotsV2Component.new(classes: "mt-4") do |component|
      component.title do
        "This is my title!"
      end
    end

    with_rendered_component_path(c) do |path|
      visit path
      assert_selector(".title", text: "This is my title!")
    end
  end
```

And the snapshot of the failed spec looks like it isn't picking up the slots.
![Captura de Pantalla 2022-11-03 a la(s) 4 58 39 p m](https://user-images.githubusercontent.com/11335191/199842582-5b4318fb-3e5f-4d90-903b-51b5dd3ec139.png)

It looks like the slots aren't being rendered

Reply:
Damn. @BlakeWilliams might you be able to have a look here?

Reply:
Ah, it looks like the example code has a mistake that's relatively common:

```ruby
c = SlotsV2Component.new(classes: "mt-4") do |component|
  component.title do
    "This is my title!"
  end
end

# needs to be 
render_inline(SlotsV2Component.new(classes: "mt-4")) do |component|
  component.title do
    "This is my title!"
  end
end
```

You can't pass blocks to `.new` because components control the initializer, and we rely on the block being passed to `render_inline` instead (same goes for `render` in views). Unfortunately you'd need `with_rendered_component_path` to use the block being passed to it to set slot information via blocks. I think we might need to rethink the API here.  Maybe something like:

```ruby
render_inline_and_visit(SlotsV2Component.new(classes: "mt-4")) do |component|
  component.title do
    "This is my title!"
  end
 end
```

Reply:
@BlakeWilliams thanks for taking a look at this. In the example you provided, as a new way to think of the API would the `render_inline_and_visit` also return a block with `path` like in the current API in this PR?

Reply:
I don't think so, since it's really "one way or the other" when it comes to being able to set slots or providing path in the block. You *could* probably combine the two in the same block but I think that might be mixing concerns and make the intention of the tests less clear so I'd likely avoid that approach.

Reply:
I think it might make sense to break away from the block concept since slots already have a use of a block. So perhaps, maybe something like:
```
path = render_inline_and_visit(SlotsV2Component.new(classes: "mt-4")) do |component|
  component.title do
    "This is my title!"
  end
end

visit path
# do your tests stuff with `path`
```

I think it makes sense perhaps to use this approach for both slots and non slot components so there isn't too many variations for how this helper is used with or without slots.

Thoughts @joelhawksley @BlakeWilliams ? Appreciate the feedback!

Reply:
I think that makes sense.

I am curious why we require consumers to call `visit` explicitly. Is there a reason for requiring consumers of the API to call `visit` explicitly instead of us calling it for them?

Reply:
> I think that makes sense.

@joelhawksley thoughts? Wanted to get your opinion on this since you mentioned the block pattern. Wondering if the approach I proposed of still works?

> I am curious why we require consumers to call visit explicitly. Is there a reason for requiring consumers of the API to call visit explicitly instead of us calling it for them?

I think was originally suggested so that we don't forcibly require someone to have capybara. From what I understand explicitly calling `visit` affords people who don't use capybara to access the component to be tested using other methods. Does that make sense?

Reply:
> I think was originally suggested so that we don't forcibly require someone to have capybara. From what I understand explicitly calling `visit` affords people who don't use capybara to access the component to be tested using other methods. Does that make sense?

I'd like to see us enumerate on that a little bit. What are some realistic scenarios where this would happen?

I imagine we could have our own "ViewComponent.interactions_test_driver` or something similar that could be used to hook into the `visit` part of `render_inline_and_visit`. If we went down that route we'd the flexibility we want (assuming we do want it) and the public API would stay lean/clear.

Reply:
> I am curious why we require consumers to call visit explicitly. Is there a reason for requiring consumers of the API to call visit explicitly instead of us calling it for them?

I guided @edwinthinks in this direction because I wanted to avoid a dependency on `visit` (and thus `Capybara`) in ViewComponent. We've previously gotten feedback and have accommodated folks that do not use `capybara`.

I don't mind having the method return a path. But then how would we clean up the temp files? The block approach allows us to use `ensure` to remove them.

Reply:
@BlakeWilliams your response got me thinking - What if we set `path` in an instance variable on the test instead of yielding it? Then we could access it in the test via an attr_accessor...

Reply:
Hmm to address this original problem with not being able to introduce slots into the current implementation. Is it possible to use `render_inline` and pass that directly to the `with_rendered_component_path`? Right now `with_rendered_component_path` is using `render_to_string` to take the HTML output and place it on a temp file.  That is, the `with_rendered_component_path` takes a component or HTML and puts it in an accessible tmp route

> I think that's fair but I'm curious how the API ergonomics differ when we focus on flexibility vs usability. Personally I feel like we should start at usability then work backwards towards flexibility. e.g. In this case maybe we could implement some kind of adapter that's hot-swappable between capybara and other solutions.

I generally agree with the usability - however I'am not sure how different it is using `visit path`. The only thing I find that could cause friction is the use of a `block` like we have now. This makes me think either yielding or the `attr_accesor` @joelhawksley you mentioned makes sense.

I auger to do a SPIKE to see how this looks without `block` syntax and support slots by refactoring the `with_rendered_component_path` to accept the output of render_inline to allow for slots.

Reply:
> @BlakeWilliams your response got me thinking - What if we set `path` in an instance variable on the test instead of yielding it? Then we could access it in the test via an attr_accessor...

I like that idea, but the only downside is that we miss having a `block` that can ensure the tmp file is deleted.

Reply:
> I like that idea, but the only downside is that we miss having a block that can ensure the tmp file is deleted.

From what I understand temp files automatically get deleted when garbage is collected, but isn't for sure to happen which can lead to leaving behind excess past temp files. Wondering though if the chances of this happening is often enough that it should influence the implementation. 

Ideally we do ensure temp files are deleted, but just wondering and thinking "outside" the box here :)

Reply:
Thanks for thinking this through @edwinthinks. Given all of these compromises, I'm 👍🏻 to having `capybara` as a dependency for this feature. All I ask is that was conditionally include this helper if `capybara` is present (or perhaps that a `visit` method is defined?) and document it as such.

Reply:
No problemo :). To verify that we are aligned are we saying we are going to breakaway from the block constant so we can avoid having to call "visit"? Moreover, I think there is still some work to be done to allow this to work with slots.

Sorry been a bit out of the loop! Might be worth catching up on a synchronous call to catch up on the direction for this PR @joelhawksley ?

Reply:
Sure thing, email me at joelhawksley@github.com with your availability for next week?

Reply:
Hey there Joel, sorry for the late reply. I'll email you with some times, not likely I can do this week or the next. Its holiday season :)